infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,275 @@
+                sessionToDelayedReplicationEventIdMap.putIfAbsent(session, new LinkedList<>());
+                // makes the value part of the map thread safe.
+                sessionToDelayedReplicationEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                    eventList.add(((LogReplicationEvent) event).getEventId());
+                    return eventList;
+                });
+            } else {
+                sessionToReplicationEventIdMap.putIfAbsent(session, new LinkedList<>());
+                // makes the value part of the map thread safe.
+                sessionToReplicationEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                    eventList.add(((LogReplicationEvent) event).getEventId());
+                    return eventList;
+                });
+            }
+        } else if (fsm.equals(FsmEventType.LogReplicationSinkEvent)){
+            LogReplicationSession session = ((LogReplicationSinkEvent) event).getSourceLeadershipManager().getSession();
+            sessionToSinkEventIdMap.putIfAbsent(session, new LinkedList<>());
+            // makes the value part of the map thread safe.
+            sessionToSinkEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                eventList.add(((LogReplicationSinkEvent) event).getEventId());
+                return eventList;
+            });
+        } else {
+            log.warn("Ignored {}. Unexpected FSM event type", fsm);
+        }
+    }
+
+    private void processSinkTask(LogReplicationSinkEvent event) {
+        LogReplicationSession session = event.getSourceLeadershipManager().getSession();
+        synchronized (event.getSourceLeadershipManager()) {
+            while (!sessionToSinkEventIdMap.get(session).get(0).equals(event.getEventId())) {
+                try {
+                    event.getSourceLeadershipManager().wait();
+                } catch (InterruptedException e) {
+                    log.error("Wait for session {} was interrupted {}", session, e.getMessage());
+                }
+            }
+        }
+
+        event.getSourceLeadershipManager().processEvent(event);
+
+        synchronized (event.getSourceLeadershipManager()) {
+            sessionToSinkEventIdMap.get(session).remove(0);
+            event.getSourceLeadershipManager().notifyAll();
+        }
+    }
+
+    private void processRuntimeTask(LogReplicationRuntimeEvent event) {
+        LogReplicationSession session = event.getRuntimeFsm().getSession();
+        // for a given session, The fsm event should be processed in the order they are queued. This condition ensures
+        // that in the event of 2 threads contending for the monitor, only the task submitted first would be processed.
+        synchronized (event.getRuntimeFsm()) {
+            while (!sessionToRuntimeEventIdMap.get(session).get(0).equals(event.getEventId())) {
+                try {
+                    event.getRuntimeFsm().wait();
```

Comment:
No, If there are say 2 events and 2 threads, the 2 threads will pick up both the events...and you can have parallel processing of 2 events for a session. We do not support this currently in our FSMs, and order is not guaranteed. (For example, when a force snapshot is triggered, we first enqueue a cancel event followed by snapshot_request event. If these events are processed out of order, we can end up cancelling the force snapshot event).

So we need to ensure that the order of event processing is the same as they were enqueued.

From the way I understand your suggestion, it will not solve this challenge..but I may be misunderstanding it. We can discuss this further if needed.

Also, currently at the most, there can be 2 outstanding events for a session. So potentially there should be only 1 thread blocked for a session. We don't have many concurrent events coming in.
There is an potential for starvation or delay in both the approaches though....but making sure that we have enough threads can help minimize it.

Reply:
Discussed offline and decided to add improvements later.